File: backend/scripts/populate_trip_template_images.py
#!/usr/bin/env python3
"""
Populate Trip Template Images Script
Automatically fetches and stores appropriate images for all trip templates
Uses multiple sources: Wikipedia, Unsplash, Mapbox static maps
"""

import os
import sys
import asyncio
import logging
from typing import Dict, Any, Optional, List
from datetime import datetime
import argparse

# Add parent directory to path for imports
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

# Supabase client
from supabase import create_client, Client
import aiohttp
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('populate_images.log'),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)

# Import the ImageService
from app.services.images.image_service import ImageService


class TripTemplateImagePopulator:
    """Populates trip templates with appropriate images from various sources"""
    
    def __init__(self, dry_run: bool = False):
        self.dry_run = dry_run
        self.supabase_url = os.getenv('SUPABASE_URL') or os.getenv('VITE_SUPABASE_URL')
        self.supabase_key = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
        
        if not self.supabase_url or not self.supabase_key:
            raise ValueError("Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY environment variables")
        
        self.supabase: Client = create_client(self.supabase_url, self.supabase_key)
        self.image_service = ImageService()
        self.stats = {
            'total': 0,
            'processed': 0,
            'updated': 0,
            'skipped': 0,
            'failed': 0
        }
        
        # Check for API keys
        self.unsplash_key = os.getenv('UNSPLASH_ACCESS_KEY')
        self.mapbox_token = os.getenv('MAPBOX_TOKEN') or os.getenv('VITE_MAPBOX_TOKEN')
        
        logger.info(f"Initialized with:")
        logger.info(f"  - Dry run: {dry_run}")
        logger.info(f"  - Unsplash API: {'Available' if self.unsplash_key else 'Not configured'}")
        logger.info(f"  - Mapbox API: {'Available' if self.mapbox_token else 'Not configured'}")
    
    async def fetch_templates_without_images(self) -> List[Dict[str, Any]]:
        """Fetch all templates that don't have images"""
        try:
            # Query templates without images
            response = self.supabase.table('trip_templates').select('*').is_('image_url', None).execute()
            
            templates = response.data
            logger.info(f"Found {len(templates)} templates without images")
            
            return templates
        except Exception as e:
            logger.error(f"Error fetching templates: {e}")
            return []
    
    async def fetch_all_templates(self) -> List[Dict[str, Any]]:
        """Fetch all templates (for forced update)"""
        try:
            response = self.supabase.table('trip_templates').select('*').execute()
            templates = response.data
            logger.info(f"Found {len(templates)} total templates")
            return templates
        except Exception as e:
            logger.error(f"Error fetching templates: {e}")
            return []
    
    def prepare_template_data(self, template: Dict[str, Any]) -> Dict[str, Any]:
        """Prepare template data for image service"""
        # Extract template_data if it exists (nested JSON)
        template_data = template.get('template_data', {})
        
        # Combine data from both levels
        prepared = {
            'name': template.get('name', ''),
            'description': template.get('description', ''),
            'category': template.get('category', 'general'),
            'region': template_data.get('region', 'Australia'),
            'highlights': template_data.get('highlights', []),
            'tags': template.get('tags', [])
        }
        
        # Parse specific template names for better image matching
        name_lower = prepared['name'].lower()
        
        # Add specific highlights based on template name
        if 'great ocean road' in name_lower:
            prepared['highlights'] = ['Twelve Apostles', 'Port Campbell', 'Lorne Beach'] + prepared['highlights']
        elif 'big lap' in name_lower:
            prepared['highlights'] = ['Australia circumnavigation', 'Uluru', 'Great Barrier Reef'] + prepared['highlights']
        elif 'east coast' in name_lower:
            prepared['highlights'] = ['Great Barrier Reef', 'Byron Bay', 'Gold Coast'] + prepared['highlights']
        elif 'red centre' in name_lower:
            prepared['highlights'] = ['Uluru', 'Kings Canyon', 'Alice Springs'] + prepared['highlights']
        elif 'tasmania' in name_lower:
            prepared['highlights'] = ['Cradle Mountain', 'Wineglass Bay', 'Port Arthur'] + prepared['highlights']
        elif 'savannah way' in name_lower:
            prepared['highlights'] = ['Kakadu National Park', 'Katherine Gorge', 'Kimberley'] + prepared['highlights']
        
        return prepared
    
    async def process_template(self, template: Dict[str, Any]) -> bool:
        """Process a single template to fetch and store images"""
        template_id = template['id']
        template_name = template['name']
        
        logger.info(f"Processing: {template_name}")
        
        try:
            # Prepare template data for image service
            template_data = self.prepare_template_data(template)
            
            # Get image using the image service
            async with self.image_service:
                image_data = await self.image_service.get_trip_template_image(template_data)
            
            if not image_data or not image_data.get('image_url'):
                logger.warning(f"  ⚠️ No image found for {template_name}")
                self.stats['failed'] += 1
                return False
            
            logger.info(f"  ✓ Found image from {image_data.get('image_source', 'unknown')}")
            logger.info(f"    Image URL: {image_data['image_url'][:50]}...")
            
            if not self.dry_run:
                # Update the database
                update_data = {
                    'image_url': image_data['image_url'],
                    'thumbnail_url': image_data.get('thumbnail_url', image_data['image_url']),
                    'updated_at': datetime.utcnow().isoformat()
                }
                
                response = self.supabase.table('trip_templates').update(update_data).eq('id', template_id).execute()
                
                if response.data:
                    logger.info(f"  ✅ Updated database for {template_name}")
                    self.stats['updated'] += 1
                    return True
                else:
                    logger.error(f"  ❌ Failed to update database for {template_name}")
                    self.stats['failed'] += 1
                    return False
            else:
                logger.info(f"  🔍 [DRY RUN] Would update {template_name}")
                self.stats['updated'] += 1
                return True
                
        except Exception as e:
            logger.error(f"  ❌ Error processing {template_name}: {e}")
            self.stats['failed'] += 1
            return False
    
    async def run(self, force_update: bool = False):
        """Run the image population process"""
        logger.info("=" * 60)
        logger.info("Starting Trip Template Image Population")
        logger.info("=" * 60)
        
        # Fetch templates
        if force_update:
            templates = await self.fetch_all_templates()
        else:
            templates = await self.fetch_templates_without_images()
        
        if not templates:
            logger.info("No templates to process")
            return
        
        self.stats['total'] = len(templates)
        
        # Process each template
        for i, template in enumerate(templates, 1):
            logger.info(f"\n[{i}/{len(templates)}] Processing template...")
            
            # Check if already has image (for non-forced updates)
            if not force_update and template.get('image_url'):
                logger.info(f"  ⏩ Skipping {template['name']} (already has image)")
                self.stats['skipped'] += 1
                continue
            
            success = await self.process_template(template)
            self.stats['processed'] += 1
            
            # Small delay to avoid rate limiting
            await asyncio.sleep(0.5)
        
        # Print summary
        logger.info("\n" + "=" * 60)
        logger.info("SUMMARY")
        logger.info("=" * 60)
        logger.info(f"Total templates: {self.stats['total']}")
        logger.info(f"Processed: {self.stats['processed']}")
        logger.info(f"Updated: {self.stats['updated']}")
        logger.info(f"Skipped: {self.stats['skipped']}")
        logger.info(f"Failed: {self.stats['failed']}")
        
        if self.dry_run:
            logger.info("\n🔍 This was a DRY RUN - no database changes were made")


async def main():
    """Main entry point"""
    parser = argparse.ArgumentParser(description='Populate trip template images')
    parser.add_argument('--dry-run', action='store_true', 
                        help='Run without updating database')
    parser.add_argument('--force', action='store_true',
                        help='Force update all templates, even those with images')
    
    args = parser.parse_args()
    
    try:
        populator = TripTemplateImagePopulator(dry_run=args.dry_run)
        await populator.run(force_update=args.force)
    except Exception as e:
        logger.error(f"Fatal error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: scripts/populate-trip-template-images.js
#!/usr/bin/env node
/**
 * Execute the existing TripTemplateImageService.ensureAllTemplatesHaveImages() method
 * This uses the built-in Wikipedia + Mapbox integration to populate all missing template images
 */

import dotenv from 'dotenv'
import { createClient } from '@supabase/supabase-js'

// Load environment variables
dotenv.config({ path: '.env.local' })

// Set up Supabase client (same as the service uses)
const supabaseUrl = process.env.VITE_SUPABASE_URL
const supabaseKey = process.env.VITE_SUPABASE_ANON_KEY

if (!supabaseUrl || !supabaseKey) {
  console.error('❌ Missing Supabase configuration')
  process.exit(1)
}

const supabase = createClient(supabaseUrl, supabaseKey)

// Inline the image population logic from TripTemplateImageService
async function populateImages() {
  try {
    console.log('🚀 Starting one-time image population for all templates...')

    // Fetch all templates without images
    const { data: templates, error } = await supabase
      .from('trip_templates')
      .select('*')
      .is('image_url', null)
      .eq('is_public', true)

    if (error) {
      console.error('Failed to fetch templates:', error)
      return
    }

    if (!templates || templates.length === 0) {
      console.log('✅ All templates already have images!')
      return
    }

    console.log(`📊 Found ${templates.length} templates without images`)

    // Wikipedia search terms for known templates (from the service)
    const wikipediaSearchTerms = {
      'aus-great-ocean-road': 'Twelve Apostles Victoria',
      'aus-big-lap': 'Uluru',
      'aus-east-coast': 'Great Barrier Reef',
      'aus-red-centre': 'Uluru Kata Tjuta National Park',
      'aus-savannah-way': 'Kakadu National Park',
      'aus-tasmania-circuit': 'Cradle Mountain',
      'aus-southwest-wa': 'Margaret River Western Australia',
      'aus-queensland-outback': 'Queensland Outback',
      'aus-murray-river': 'Murray River Australia',
      'aus-gibb-river-road': 'Gibb River Road',
      'aus-high-country': 'Victorian Alps',
      'aus-nullarbor': 'Nullarbor Plain',
      'aus-cape-york': 'Cape York Peninsula',
      'aus-flinders-ranges': 'Flinders Ranges',
      'aus-sunshine-coast': 'Sunshine Coast Queensland',
    }

    // Process templates in batches
    const batchSize = 3
    let successCount = 0
    let mapboxCount = 0

    for (let i = 0; i < templates.length; i += batchSize) {
      const batch = templates.slice(i, i + batchSize)

      await Promise.all(
        batch.map(async (dbTemplate) => {
          const template = {
            id: dbTemplate.id,
            name: dbTemplate.name,
            region: dbTemplate.template_data?.region || 'Australia',
            category: dbTemplate.category
          }

          // Try Wikipedia first if we have a search term
          const searchTerm = wikipediaSearchTerms[template.id]
          let success = false

          if (searchTerm) {
            success = await fetchAndStoreWikipediaImage(template, searchTerm)
            if (success) successCount++
          }

          // Fallback to Mapbox if Wikipedia failed
          if (!success) {
            success = generateAndStoreMapboxImage(template)
            if (success) mapboxCount++
          }
        })
      )

      // Small delay between batches to be respectful to APIs
      if (i + batchSize < templates.length) {
        await new Promise(resolve => setTimeout(resolve, 500))
      }
    }

    console.log(`✅ Image population complete!`)
    console.log(`   - Wikipedia images: ${successCount}`)
    console.log(`   - Mapbox fallbacks: ${mapboxCount}`)
    console.log(`   - Total processed: ${templates.length}`)

  } catch (error) {
    console.error('💥 Failed to populate template images:', error)
  }
}

// Helper functions (simplified versions from the service)
async function fetchAndStoreWikipediaImage(template, searchTerm) {
  try {
    console.log(`🔍 Fetching Wikipedia image for ${template.name} using term: ${searchTerm}`)

    // Fetch from Wikipedia API
    const response = await fetch(
      `https://en.wikipedia.org/api/rest_v1/page/summary/${encodeURIComponent(searchTerm)}`
    )

    if (!response.ok) {
      console.warn(`Wikipedia page not found for: ${searchTerm}`)
      return false
    }

    const data = await response.json()

    // Extract image URLs
    let imageUrl = null
    let thumbnailUrl = null

    if (data.thumbnail?.source) {
      thumbnailUrl = data.thumbnail.source
      // Get higher resolution version for main image
      imageUrl = data.thumbnail.source.replace(/\/\d+px-/, '/800px-')
    } else if (data.originalimage?.source) {
      imageUrl = data.originalimage.source
    }

    if (!imageUrl) {
      console.warn(`No image found on Wikipedia for: ${searchTerm}`)
      return false
    }

    // Store in database
    return await updateTemplateImage(
      template.id,
      imageUrl,
      thumbnailUrl,
      'wikipedia',
      `Wikipedia - ${searchTerm}`
    )
  } catch (error) {
    console.error(`Error fetching Wikipedia image for ${searchTerm}:`, error)
    return false
  }
}

function generateAndStoreMapboxImage(template) {
  const mapboxToken = process.env.VITE_MAPBOX_TOKEN
  if (!mapboxToken) {
    console.warn('No Mapbox token available for map generation')
    return false
  }

  // Determine map center based on template data
  let centerLat = -25.2744 // Default to Australia
  let centerLon = 133.7751
  let zoom = 4

  // Region-specific coordinates
  const regionCoords = {
    'Australia': [-25.2744, 133.7751, 4],
    'Victoria': [-37.4713, 144.7852, 6],
    'Queensland': [-20.9176, 142.7028, 5],
    'New South Wales': [-31.8401, 145.6121, 5],
    'Tasmania': [-41.4545, 145.9707, 6],
    'Western Australia': [-27.6728, 121.6283, 5],
  }

  // Use region-specific coordinates if available
  if (template.region && regionCoords[template.region]) {
    [centerLat, centerLon, zoom] = regionCoords[template.region]
  }

  // Generate static map URL
  const marker = `pin-l-star+3b82f6(${centerLon},${centerLat})`
  const imageUrl = `https://api.mapbox.com/styles/v1/mapbox/outdoors-v12/static/${marker}/${centerLon},${centerLat},${zoom},0/800x400@2x?access_token=${mapboxToken}`
  const thumbnailUrl = `https://api.mapbox.com/styles/v1/mapbox/outdoors-v12/static/${marker}/${centerLon},${centerLat},${zoom},0/400x200@2x?access_token=${mapboxToken}`

  // Store in database (async but we don't wait)
  updateTemplateImage(
    template.id,
    imageUrl,
    thumbnailUrl,
    'mapbox',
    'Mapbox Static API'
  ).catch(error => {
    console.error(`Failed to store Mapbox image for ${template.id}:`, error)
  })

  return true
}

async function updateTemplateImage(templateId, imageUrl, thumbnailUrl, source, attribution) {
  try {
    console.log(`📸 Updating image for template ${templateId}`)

    const updateData = {
      image_url: imageUrl,
      image_source: source || 'manual'
    }

    if (thumbnailUrl) {
      updateData.thumbnail_url = thumbnailUrl
    }

    if (attribution) {
      updateData.image_attribution = attribution
    }

    const { error } = await supabase
      .from('trip_templates')
      .update(updateData)
      .eq('id', templateId)

    if (error) {
      console.error(`❌ Failed to update image for template ${templateId}:`, error)
      return false
    }

    console.log(`✅ Successfully updated image for template ${templateId}`)
    return true
  } catch (error) {
    console.error(`💥 Unexpected error updating template image:`, error)
    return false
  }
}

// Run the script
populateImages()